### PR TITLE
cache: add SetSnapshot and GetSnapshot to LinearCache

### PR DESCRIPTION
**Pull request by jakubdyszkiewicz**
The implementation discussed in #433

`SetSnapshot` and `GetSnapshot` enables users to use `LinearCache` in a more convenient way if the major use case is sending `DiscoveryRequest` with empty `resourceNames`.

### CODE REVIEW

**Comment by jakubdyszkiewicz** [comment]
Sorry about the force push 🙏  I started the work from `master` instead of `main`.

**Comment by alecholmez** [comment]
Thanks for opening this! I'll review

File: pkg/cache/v3/linear.go [hunk 1]
```diff
@@ -239,6 +239,45 @@ func (cache *LinearCache) CreateWatch(request *Request) (chan Response, func())
 	}
 }
 
+// SetResources replaces current resources with a new set of resources
+//
+// If we have individual watches for two resources (there are 2 DiscoveryRequest with resourceNames=res1 and resourceNames=res2)
+// Then if you set new resources, res1 changed but res2 is not, we assume that all resources are changed so both watches will be triggered.
+// That's why this function is well suited for a use case when resourceNames of DiscoveryRequest is empty.
+// This way you don't have to call UpdateResource for every resource which will trigger watch of DiscoveryRequest(resourceNames="") twice,
+// but it will only happen once regardless of how many resources are changed.
```
Comment (alecholmez) [inline pkg/cache/v3/linear.go:248]: I think this comment might be a little much for this function. If I understand this correctly, this is a good hook to use wildcarding when requesting resources (resourceName being empty), maybe we should just allude to that here instead?

File: pkg/cache/v3/linear.go [hunk 2]
```diff
@@ -239,6 +239,45 @@ func (cache *LinearCache) CreateWatch(request *Request) (chan Response, func())
...
+	cache.notifyAll(modified)
+}
+
+// GetResources returns current resources stored in the cache
+func (cache *LinearCache) GetResources() map[string]types.Resource {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+	return cache.resources
+}
+
```
Comment (alecholmez) [inline pkg/cache/v3/linear.go:280]: Just a nit pick but can we move both these functions above the `CreateWatch()`? I'd like to keep them bundled with Update/DeleteResource().

File: pkg/cache/v3/linear.go [hunk 3]
```diff
@@ -239,6 +239,45 @@ func (cache *LinearCache) CreateWatch(request *Request) (chan Response, func())
...
+		cache.versionVector[name] = cache.version
+		modified[name] = struct{}{}
+	}
+
+	cache.notifyAll(modified)
+}
+
+// GetResources returns current resources stored in the cache
+func (cache *LinearCache) GetResources() map[string]types.Resource {
+	cache.mu.Lock()
```
Comment (alecholmez) [inline pkg/cache/v3/linear.go:276]: This is just a read right? Could we change this to an `RLock()`?

**Review by alecholmez — CHANGES_REQUESTED** [review]
Just a few things so far

File: pkg/cache/v3/linear.go (hunk 1 above)
Comment (jakubdyszkiewicz) [reply pkg/cache/v3/linear.go]: I rewrote the comment. Please take a look if it's ok, if not I could use some guidance. Thanks!

File: pkg/cache/v3/linear.go (hunk 3 above)
Comment (jakubdyszkiewicz) [reply pkg/cache/v3/linear.go]: Right, changed to RWLock and changed `NumWatces` to also only use `RLock()`

**Comment by jakubdyszkiewicz** [comment]
hey, @alecholmez do you have any more suggestions?

**Comment by alecholmez** [comment]
@jakubdyszkiewicz I think this is good from my end, can you address the merge conflicts?

@snowp you have any thoughts?

File: pkg/cache/v3/linear.go [hunk 4]
```diff
@@ -165,6 +165,41 @@ func (cache *LinearCache) DeleteResource(name string) error {
...
+// This function is useful for use cases when DiscoveryRequest#resourceNames is empty.
+// This way watches that follow all resources are triggered only once regardless of how many resources are changed.
+func (cache *LinearCache) SetResources(resources map[string]types.Resource) {
+	cache.mu.Lock()
+	defer cache.mu.Unlock()
+
+	cache.version += 1
+
+	modified := map[string]struct{}{}
+	for name := range cache.resources {
```
Comment (snowp) [inline pkg/cache/v3/linear.go:178]: Some comments would be make this easier to follow, like `// First account for removed resources.` and a similar one for the other for loop

File: pkg/cache/v3/linear.go [hunk 5]
```diff
@@ -165,6 +165,41 @@ func (cache *LinearCache) DeleteResource(name string) error {
...
+		if _, found := resources[name]; !found {
+			delete(cache.versionVector, name)
+			modified[name] = struct{}{}
+		}
+	}
+
+	cache.resources = resources
+	for name := range resources {
+		// We assume all resources passed to SetResources are changed.
+		// Otherwise we would have to do proto.Equal on resources which is pretty expensive operation
```
Comment (snowp) [inline pkg/cache/v3/linear.go:188]: There are other solutions to this (like per version resource values like how Delta is modeled), but it's probably fine to leave it as this for now.

File: pkg/cache/v3/linear_test.go [hunk 6]
```diff
@@ -129,6 +130,64 @@ func TestLinearBasic(t *testing.T) {
 	verifyResponse(t, w, "3", 2)
 }
 
+func TestLinearSetResources(t *testing.T) {
+	c := NewLinearCache(testType)
+
+	// Create new resources
+	w1, _ := c.CreateWatch(&Request{ResourceNames: []string{"a"}, TypeUrl: testType, VersionInfo: "0"})
+	mustBlock(t, w1)
+	w, _ := c.CreateWatch(&Request{TypeUrl: testType, VersionInfo: "0"})
```
Comment (snowp) [inline pkg/cache/v3/linear_test.go:139]: nit: name this variable w2 for consistency

**Review by snowp — COMMENTED** [review]
Thanks this looks good to me for the most part, just a few minor nits

**Comment by jakubdyszkiewicz** [comment]
@snowp fixed comments and conflicts

**Review by snowp — APPROVED** [review]
Thanks!

**Comment by snowp** [comment]
@alecholmez for merge

**Review by alecholmez — APPROVED** [review]
This looks good @jakubdyszkiewicz, thanks for the contribution!

**Pull request merged**